lib/slave/heartbeat.go

```diff
@@ -12,24 +12,43 @@ import (
 
 func (s *Slave) sendHeartbeat() {
 	log.Println("[Heartbeat] Sending...")
-	_, err := postHeartbeat(s)
+	heartbeat := s.generateHeartbeat()
+	_, err := s.postHeartbeat(heartbeat)
 
-	if err == nil {
-		var taskStatusses []lib.TaskStatus
-		for _, ts := range s.heartbeat.TaskStatus {
-			if ts.Status == lib.Running {
-				taskStatusses = append(taskStatusses, ts)
+	if err != nil {
+		log.Println("[Heartbeat] Delivery failed")
+	}
+}
+
+func (s *Slave) generateHeartbeat() lib.Heartbeat {
+	heartbeat := lib.Heartbeat{
+		SlaveId: s.id,
+	}
+	for i, task := range s.tasks {
+		var progress string
+		if task.Status == lib.Running {
+			c := make(chan string)
+			s.tasks[i].progressChan <- c
+			select {
+			case progress = <-c:
+			case <-time.After(time.Second * 1): // This could accually happen for a legitimate reason (much hash interations, slow hashing algorithms), but for now see what happens
```

Comment:
Interesting, if we're using a lot of iterations then this may indeed happen. A possible fix would be to use mutex instead of channels to communicate the pgoress. But let's see what happens first...